### DISCUSSION TITLE
How to control these macro definitions in the shader?

### DISCUSSION THREAD

**Discussion by dugufly** [google/filament#4621] [Q&A]
If I use UbershaderLoader, how to control these macro definitions in the shader: HAS_ATTRIBUTE_TANGENTS， HAS_ ATTRIBUTE_ TANGENTS， MATERIAL_HAS_SHEEN_COLOR,  IBL_INTEGRATION？

**Comment by romainguy** [comment]
You don't control those directly, they are inferred from what you do in your material. For instance if you specify a `material.sheenColor, `MATERIAL_HAS_SHEEN_COLOR` will be set. Those are implementation details you don't need to worry about. Everything public is documented in https://google.github.io/filament/Materials.html

**Reply by dugufly** [reply]
Where is this correspondence implemented？ 

> “For instance if you specify a material.sheenColor, MATERIAL_HAS_SHEEN_COLOR` will be set”
